kitsune-fe/src/components/Captcha.vue

```diff
@@ -0,0 +1,84 @@
+<template>
+  <FormKit
+    type="hidden"
+    name="captchaToken"
+    validation="required"
+    v-model="captchaState.token"
+  />
+  <vue-hcaptcha
+    v-if="backend == CaptchaBackend.HCaptcha"
+    :sitekey="sitekey"
+    @verify="onVerify"
+    @expired="onExpire"
+    @error="onError"
+    @challenge-expired="onExpire"
+  >
+  </vue-hcaptcha>
+  <div v-if="backend === CaptchaBackend.MCaptcha">
+    <div id="mcaptcha__widget-container"></div>
+    <component
+      is="script"
+      type="text/javascript"
+      src="/public/mcaptcha-glue.js"
+      onload="mcaptcha__load();"
+    ></component>
+    <component is="script" type="text/javascript" charset="utf-8">
+      function mcaptcha__load() { let config = { widgetLink: new URL("{{
+        sitekey
+      }}"), }; new mcaptchaGlue.default(config); }
+    </component>
+  </div>
+</template>
+
+<script setup lang="ts">
+  import VueHcaptcha from '@hcaptcha/vue3-hcaptcha';
+
+  import { reactive, onMounted } from 'vue';
+
+  import { CaptchaBackend } from '../graphql/types/graphql';
+
+  defineProps<{
+    backend: CaptchaBackend;
+    sitekey: string;
+  }>();
+
+  const captchaState = reactive({
+    verified: false,
+    expired: false,
+    token: '',
+    error: '',
+  });
+
+  function onVerify(tokenStr: string) {
+    captchaState.verified = true;
+    captchaState.token = tokenStr;
+  }
+
+  function onExpire() {
+    captchaState.verified = false;
+    captchaState.token = '';
+    captchaState.expired = true;
+  }
+
+  function onError(err: string) {
+    captchaState.token = '';
+    captchaState.error = err;
+  }
+
+  // this is the only to capture mCaptcha token
+  onMounted(() => {
+    window.addEventListener('message', (e) => {
```

Comment:
I'm pretty sure our linter configuration will complain about this.

Check the output of `yarn lint`

Reply:
It did not, actually. I fixed the other linter complaints though, except for one.
If you apply this suggestion:
```
/src/components/AuthForms.vue
  99:22  warning  Replace `username:·$username,·email:·$email,·password:·$password,·captchaToken:·$captchaToken` with `⏎··········username:·$username⏎··········email:·$email⏎··········password:·$password⏎··········captchaToken:·$captchaToken⏎········`  prettier/prettier
```
You get an error:
```
src/components/AuthForms.vue:92:5 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'string | DocumentNode | TypedDocumentNode<any, AnyVariables>'.

 92     graphql(`
        ~~~~~~~~~
 93       mutation registerUser(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
107       }
    ~~~~~~~
108     `),
    ~~~~~~
```

Reply:
It uses an interesting strategy of selecting generated TypeScript types by indexing a map for the correct type, where the GraphQL query is the key.  
That's why you can't change the query without regenerating the code.